### PR TITLE
Spdz mac check issues

### LINKED ISSUE #152: Check the spdz maccheck protocol
Make sure the mac check protocol of spdz is used correctly

### PR DESCRIPTION
**Pull request by Rhaea**
This fixes #152 and #34. We no longer cut corners, but actually check for cheating before and after output is made.

### CODE REVIEW

**Comment by Rhaea** [comment]
@pffrandsen I agree that indeed we could do a Mac check before outputs and then after outputs as long as it is in the same batch. This holds true. However, in order for me not to wreck the native protocol round concept, I need the output protocols to not be evaluated by the evaluator, but only by SPDZ itself. This is currently not supported, and is so far pretty specific to SPDZ. Therefore I did not change the design. 
I agree that prettier ways to do this is possible, but would require design changes that we need to ensure are not just there for SPDZ only. Most likely other protocol suites using the MAC style could benefit though which includes TinyTables once active security is there. 
I'll wait for the other Peter to respond as well before making design changes.

File: suite/spdz/src/main/java/dk/alexandra/fresco/suite/spdz/SpdzBatchedStrategy.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+  /**
+   * @param protocols array holding the protocols to be evaluated
+   * @param sceNetwork array of sceNetworks corresponding to the protocols to be evaluated. I.e.,
+   *        the array should contain numProtocols SCENetworks, with sceNetwork[i] used for
+   *        communication in protocols[i].
+   * @param channel string indicating the channel to communicate over.
+   * @param rp the resource pool.
+   * @param startingRound The round to start from
+   */
```
Comment (GuutBoy) [inline suite/spdz/src/main/java/dk/alexandra/fresco/suite/spdz/SpdzBatchedStrategy.java]: The documentation is not in sync with the method.

File: suite/spdz/src/main/java/dk/alexandra/fresco/suite/spdz/SpdzRoundSynchronization.java [hunk 2]
```diff
@@ -4,55 +4,70 @@
...
  */
 public class SpdzRoundSynchronization implements RoundSynchronization<SpdzResourcePool> {
 
   private static final int macCheckThreshold = 100000;
+  private int gatesEvaluated = 0;
 
-  private void MACCheck(SpdzResourcePool resourcePool,
-      SCENetwork sceNetworks) throws IOException {
-
+  private void doMACCheck(SpdzResourcePool resourcePool, SCENetwork sceNetworks)
```
Comment (GuutBoy) [inline suite/spdz/src/main/java/dk/alexandra/fresco/suite/spdz/SpdzRoundSynchronization.java]: sceNetworks -> sceNetwork

File: suite/spdz/src/main/java/dk/alexandra/fresco/suite/spdz/datatypes/SpdzOutputProtocol.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package dk.alexandra.fresco.suite.spdz.datatypes;
```
Comment (GuutBoy) [inline suite/spdz/src/main/java/dk/alexandra/fresco/suite/spdz/datatypes/SpdzOutputProtocol.java]: Shouldn't this by in spdz.gates?

**Comment by pffrandsen** [comment]
The minor of the two issues has been fixed.

The major has three resolutions;
- We keep the current hack but change the rest of the code to match the new interface. I would be happy to share the very long list of entailed changes if we redefine the NativeProtocol interface. 
- We rework the current solution to use and work with round synchronization
- We live with the current hack

The latter is unacceptable in my opinion

**Comment by GuutBoy** [comment]
Hej Peter,

Det kan der være noget om. Jeg satte den bare til default.

Jeg har været syg i går og er stadig meget dårlig. Jeg vil kigge på det når jeg er tilbage.

/Peter

**Comment by Rhaea** [comment]
I actually liked the beforeBatch idea once I started pondering it. It makes it possible also as a protocol suite to inject or even remove protocols within the batch. In this case, we just check if an output gate exists, and then do a MAC check on all previous layers. We then finish the batch with another MAC check if an output gate existed in the current batch. 
I removed the hack.
All should be good now from my point of view.

File: suite/spdz/src/main/java/dk/alexandra/fresco/suite/spdz/gates/SpdzOutputProtocol.java [hunk 4]
```diff
@@ -1,95 +1,13 @@
...
-  }
+/**
+ * Marks a protocol as being able to output to some party. This means that the protocol should be
+ * added to the list of output protocols in the current batch. It also means that calling eval
+ * starting from round 1 will trigger the actual evaluation of the output protocol.
+ * 
+ * @author Kasper Damgaard
+ *
+ */
+public abstract class SpdzOutputProtocol<OutputT> extends SpdzNativeProtocol<OutputT> {
```
Comment (pffrandsen) [inline suite/spdz/src/main/java/dk/alexandra/fresco/suite/spdz/gates/SpdzOutputProtocol.java]: This should be an interface - it is only used as such

**Comment by GuutBoy** [comment]
@Rhaea Would it be possible for you to fix the few conflicts here so we can get this merged?

**Comment by Rhaea** [comment]
Interleaving would theoretically break security as the invariant is that the tree from an output gate is MAC checked before output is delivered. If we now included the check in the next batch, this would mean that output is delivered potentially before the mac check is completed. Whether this happens or not depends on the evaluation strategy. 
But this does not mean there is not a more efficient solution. So sure, we can look at that at some later point yes.

**Pull request merged**